src/Hosting/TestHost/test/WebSocketClientTests.cs

```diff
@@ -54,5 +55,77 @@ public async Task ConnectAsync_ShouldSetRequestProperties(string requestUri, str
             Assert.Equal(expectedHost, capturedHost);
             Assert.Equal("/connect", capturedPath);
         }
+
+        [Fact]
+        public async Task CanAcceptWebSocket()
+        {
+            using (var testServer = new TestServer(new WebHostBuilder()
+                .Configure(app =>
+                {
+                    app.UseWebSockets();
+                    app.Run(async ctx =>
+                    {
+                        if (ctx.Request.Path.StartsWithSegments("/connect"))
+                        {
+                            if (ctx.WebSockets.IsWebSocketRequest)
+                            {
+                                using var websocket = await ctx.WebSockets.AcceptWebSocketAsync();
+                                var buffer = new byte[1000];
+                                var res = await websocket.ReceiveAsync(buffer, default);
+                                await websocket.SendAsync(buffer.AsMemory(0, res.Count), System.Net.WebSockets.WebSocketMessageType.Binary, true, default);
+                                await websocket.CloseAsync(System.Net.WebSockets.WebSocketCloseStatus.NormalClosure, null, default);
+                            }
+                        }
+                    });
+                })))
+            {
+                var client = testServer.CreateWebSocketClient();
+
+                using var socket = await client.ConnectAsync(
+                    uri: new Uri("http://localhost/connect"),
+                    cancellationToken: default);
+
+                await socket.SendAsync(new byte[10], System.Net.WebSockets.WebSocketMessageType.Binary, true, default);
+                var res = await socket.ReceiveAsync(new byte[100], default);
+                Assert.Equal(10, res.Count);
+                Assert.True(res.EndOfMessage);
+
+                await socket.CloseAsync(System.Net.WebSockets.WebSocketCloseStatus.NormalClosure, null, default);
+            }
+        }
+
+        [Fact]
+        public async Task VerifyWebSocketAndUpgradeFeatures()
+        {
+            using (var testServer = new TestServer(new WebHostBuilder()
+                .Configure(app =>
+                {
+                    app.Run(async c =>
+                    {
+                        var upgradeFeature = c.Features.Get<IHttpUpgradeFeature>();
+                        Assert.NotNull(upgradeFeature);
+                        Assert.False(upgradeFeature.IsUpgradableRequest);
+                        await Assert.ThrowsAsync<NotSupportedException>(() => upgradeFeature.UpgradeAsync());
+
+                        var webSocketFeature = c.Features.Get<IHttpWebSocketFeature>();
+                        Assert.NotNull(webSocketFeature);
+                        Assert.True(webSocketFeature.IsWebSocketRequest);
+                    });
+                })))
+            {
+                var client = testServer.CreateWebSocketClient();
```

Comment:
Sorry, I meant that you'd send a non-websocket request here like you do in the signalr negotiation scenario. Including `app.UseWebSockets();` in this test makes sense and the checking that it does add the `IHttpWebSocketFeature` SignalR needs.